validator/lib/jsonapi/validator/resource.rb

```diff
@@ -0,0 +1,123 @@
+require 'jsonapi/validator/document'
+require 'jsonapi/validator/relationship'
+
+module JSONAPI
+  module Validator
+    class Resource
```

Comment:
Not that this is also motivated by the fact that when deserializing to an intermediate structure, you are usually left with doing the exact same validation work for building your custom structures as you would have done directly on the payload itself.